### PR TITLE
HPCC-32250 Track PeakTempSize for all activities that use lookahead

### PR DESCRIPTION
**Pull request by shamser**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

### CODE REVIEW

File: thorlcr/activities/enth/thenth.cpp [hunk 1]
```diff
@@ -37,7 +37,7 @@ class CEnthActivityMaster : public CMasterActivity
 CActivityBase *createEnthActivityMaster(CMasterGraphElement *container)
 {
     if (container->queryLocalOrGrouped())
-        return new CMasterActivity(container);
+        return new CMasterActivity(container, spillingActivityStatistics);
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply thorlcr/activities/enth/thenth.cpp:40]: in fact - both local and global Enth *may* use a lookahead, but I don't think localenth should - I will open a separate JIRA to change it - for now it is better to mirror how it's implemented and ensure spillingActivityStatistics used for both (as it is in worker impl.)

File: thorlcr/activities/selectnth/thselectnth.cpp [hunk 2]
```diff
@@ -35,7 +35,7 @@ class CSelectNthActivityMaster : public CMasterActivity
 CActivityBase *createSelectNthActivityMaster(CMasterGraphElement *container)
 {
     if (container->queryLocalOrGrouped())
-        return new CMasterActivity(container);
+        return new CMasterActivity(container, spillingActivityStatistics);
```
Comment (jakesmith) [inline thorlcr/activities/selectnth/thselectnth.cpp:38]: as per previous comments. The local version doesn't use a lookahead. The global does.

**Review by jakesmith — CHANGES_REQUESTED** [review]
@shamser - please see comments.

I do wonder whether we shouldn't just add spilling stats to the base activity stats.
The downside, mergeStats will walk over stats in the mapping, and query them, in activities that never used them, unnecessarily.
The upside, cleaner code, future changes won't have to keep these common attributes in sync.

@ghalliday - what's your thoughts? is the overhead of cycling through a few unused stat. kinds an issue. NB: in general there will be many stats that are are used I suspect. We could also think about optimizing mergeStats/tracking the stats that have been used..

File: thorlcr/activities/choosesets/thchoosesets.cpp [hunk 3]
```diff
@@ -34,7 +34,7 @@ class CChooseSetsActivityMaster : public CMasterActivity
 CActivityBase *createChooseSetsActivityMaster(CMasterGraphElement *container)
 {
     if (container->queryLocalOrGrouped())
-        return new CMasterActivity(container);
+        return new CMasterActivity(container, spillingActivityStatistics);
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply thorlcr/activities/choosesets/thchoosesets.cpp:37]: Doesn't local use lookahead?  Simpler for all use use spilling stats.

File: thorlcr/activities/choosesets/thchoosesetsslave.cpp [hunk 4]
```diff
@@ -30,7 +30,7 @@ class BaseChooseSetsActivity : public CSlaveActivity
     unsigned *tallies;
 
 public:
-    BaseChooseSetsActivity(CGraphElementBase *_container) : CSlaveActivity(_container)
+    BaseChooseSetsActivity(CGraphElementBase *_container, const StatisticsMapping & _mapping = basicActivityStatistics) : CSlaveActivity(_container, _mapping)
```
Comment (ghalliday) [inline thorlcr/activities/choosesets/thchoosesetsslave.cpp:33]: Clearer/safer to not have a default parameter so it is easy to check the correct mapping is being used for all variants.

File: thorlcr/activities/firstn/thfirstnslave.cpp [hunk 5]
```diff
@@ -35,7 +35,7 @@ class CFirstNSlaveBase : public CSlaveActivity
     IHThorFirstNArg *helper;
 
 public:
-    CFirstNSlaveBase(CGraphElementBase *_container) : CSlaveActivity(_container)
+    CFirstNSlaveBase(CGraphElementBase *_container, const StatisticsMapping & _mapping = basicActivityStatistics) : CSlaveActivity(_container, _mapping)
```
Comment (ghalliday) [inline thorlcr/activities/firstn/thfirstnslave.cpp:38]: similar comment to above about not using a default parameter

File: thorlcr/thorutil/thormisc.cpp [hunk 6]
```diff
@@ -79,15 +79,15 @@ const StatisticsMapping basicActivityStatistics({StTimeTotalExecute, StTimeLocal
 const StatisticsMapping groupActivityStatistics({StNumGroups, StNumGroupMax}, basicActivityStatistics);
 const StatisticsMapping indexReadFileStatistics({}, diskReadRemoteStatistics, jhtreeCacheStatistics);
 const StatisticsMapping indexReadActivityStatistics({StNumRowsProcessed}, indexReadFileStatistics, basicActivityStatistics);
-const StatisticsMapping indexWriteActivityStatistics({StPerReplicated, StNumLeafCacheAdds, StNumNodeCacheAdds, StNumBlobCacheAdds }, basicActivityStatistics, diskWriteRemoteStatistics);
+const StatisticsMapping indexWriteActivityStatistics({StPerReplicated, StNumLeafCacheAdds, StNumNodeCacheAdds, StNumBlobCacheAdds }, spillingActivityStatistics, basicActivityStatistics, diskWriteRemoteStatistics);
```
Comment (ghalliday) [inline thorlcr/thorutil/thormisc.cpp:82]: spillingActivityStatistics -> spillStatistics

**Review by ghalliday — COMMENTED** [review]
@jakesmith I agree with your comments.

**Comment by ghalliday** [comment]
Replaced by https://github.com/hpcc-systems/HPCC-Platform/pull/18920

**Pull request closed without merging**